### PR TITLE
[ctx_prof] API to get the instrumentation of a BB

### PR DESCRIPTION
**Pull request by mtrofin**
Analogous to PR #104491 

Issue #89287

### CODE REVIEW

**Comment by mtrofin** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/105468?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#105469** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/105469?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#105468** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/105468?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#104539** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/104539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#104491** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/104491?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#104490** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/104490?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Analysis/CtxProfAnalysis.cpp [hunk 1]
```diff
@@ -202,6 +202,13 @@ InstrProfCallsite *CtxProfAnalysis::getCallsiteInstrumentation(CallBase &CB) {
   return nullptr;
 }
 
+InstrProfIncrementInst *CtxProfAnalysis::getBBInstrumentation(BasicBlock &BB) {
+  for (auto &I : BB)
```
Comment (boomanaiden154) [inline llvm/lib/Analysis/CtxProfAnalysis.cpp:206]: I think explicitly typing `Instruction` would improve readability a bit.

File: llvm/unittests/Analysis/CtxProfAnalysisTest.cpp [hunk 2]
```diff
@@ -132,4 +132,26 @@ TEST_F(CtxProfAnalysisTest, GetCallsiteIDNegativeTest) {
   EXPECT_EQ(IndIns, nullptr);
 }
 
+TEST_F(CtxProfAnalysisTest, GetBBIDTest) {
+  ModulePassManager MPM;
+  MPM.addPass(PGOInstrumentationGen(PGOInstrumentationType::CTXPROF));
+  EXPECT_FALSE(MPM.run(*M, MAM).areAllPreserved());
+  auto *F = M->getFunction("foo");
+  ASSERT_NE(F, nullptr);
+  std::map<std::string, int> BBNameAndID;
```
Comment (boomanaiden154) [inline llvm/unittests/Analysis/CtxProfAnalysisTest.cpp:141]: It's a test and doesn't matter much, but should we more carefully choose the container here? It doesn't seem like the iteration order matters, so something like `std::unordered_map` might be better.

File: llvm/unittests/Analysis/CtxProfAnalysisTest.cpp [hunk 3]
```diff
@@ -132,4 +132,26 @@ TEST_F(CtxProfAnalysisTest, GetCallsiteIDNegativeTest) {
...
+
+  for (auto &BB : *F) {
+    auto *Ins = CtxProfAnalysis::getBBInstrumentation(BB);
```
Comment (boomanaiden154) [inline llvm/unittests/Analysis/CtxProfAnalysisTest.cpp:144]: Explicitly type this instruction?

**Review by boomanaiden154 — APPROVED** [review]
Nits, otherwise LGTM.

File: llvm/lib/Analysis/CtxProfAnalysis.cpp (hunk 1 above)
Comment (mtrofin) [reply llvm/lib/Analysis/CtxProfAnalysis.cpp:206]: why, this is a typical iteration pattern for BBs (doing a very naive search though llvm/lib for either patterns - i.e. `for (auto &I : BB)` or `for (Instruction &I : BB)` reveals about the same nr of instances.

**Comment by mtrofin** [comment]
### Merge activity

* **Aug 21, 8:16 PM EDT**: @mtrofin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/105468).

**Pull request merged**